_posts/hacks/2017-04-14-How to sign.md

```diff
@@ -0,0 +1,208 @@
+                     new JcaDigestCalculatorProviderBuilder().build())
+                     .build(sha256Signer, signCert));
+
+      gen.addCertificates(certs);
+
+      CMSSignedData sigData = gen.generate(msg, false);
+```
+
+Note, that you can define, if also the data and not only signature should be in the CMS-Container.
+With this container you got the signature,
+the certificate, that can be used for verifying, digital algorithm, probably the signed data etc.
+It is also possible to create several signatures for the data and put them in the container. It means you can send this kind of signature
+to a third party, for instance, you can sign data, that are provided by a web service and the consumer wants to verify the signature.
+The code to verify a *CMSSignedData* (again JavaDoc of BounceCastle)
+
+```java
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cms.CMSException;
+import org.bouncycastle.cms.CMSSignedData;
+import org.bouncycastle.cms.SignerInformation;
+import org.bouncycastle.cms.SignerInformationStore;
+import org.bouncycastle.cms.jcajce.JcaSimpleSignerInfoVerifierBuilder;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.util.Store;
+
+Store certStore = cmsSignedData.getCertificates();
+SignerInformationStore signers = cmsSignedData.getSignerInfos();
+Collection c = signers.getSigners();
+Iterator it = c.iterator();
+
+while (it.hasNext())
+  {
+    SignerInformation signer = (SignerInformation)it.next();
+    Collection certCollection = certStore.getMatches(signer.getSID());
+    Iterator certIt = certCollection.iterator();
+    X509CertificateHolder cert = (X509CertificateHolder)certIt.next();
+    if (signer.verify(new JcaSimpleSignerInfoVerifierBuilder().build(cert))) {
+           verified++;
+       }
+   }
+```
+
+If you want to use the whole functionality of a JCE implementation you have to install unrestricted policy files for the JVM. If you don't,
+you'll get something like this
+```java
+java.lang.SecurityException: Unsupported keysize or algorithm parameters" or "java.security.InvalidKeyException: Illegal key size
+```
+
+As you guess it is not a big problem to get and install the unrestricted policy files for *your* JVM. But what if you 
+want to distribute you application. It can be pretty difficult for some users to solve this problem. The BouncyCastle library
+has a solution for this problem. It provides light weight version of cryptographic operation. It means, that these operations don't use any JCE
+provider. That's why it is not necessary to install unrestricted policy files. Maybe you already saw, that some classes of the BouncyCastle
+begin with _Jce_  (Java Cryptography Extension) or with _Jca_(Java Cryptography Architecture). These classes use JCE-Provider.
+The light weight classes begin with _Bc_.
+The code for signing with light weight version would look like this
+
+```java
+import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
```

Comment:
The lines of the code blocks are too long. So you need to scroll in the sources: take a look at the preview:
https://deploy-preview-51--reflectoring.netlify.com/hacks/how-to-sign/